src/webportal/src/app/user/user-view/user-view.component.js

```diff
@@ -250,6 +258,174 @@ const updateUserGithubPAT = (username) => {
+    [columnVC]: 'default',
+    [columnGithubPAT]: '',
+  }]);
+  let universalBOM = '\uFEFF';
+  let filename = 'userinfo.csv';
+  let file = new Blob([universalBOM + csvString], {type: 'text/csv;charset=utf-8'});
+  if (window.navigator.msSaveOrOpenBlob) { // IE10+
+    window.navigator.msSaveOrOpenBlob(file, filename);
+  } else { // Others
+    let a = document.createElement('a');
+    let url = URL.createObjectURL(file);
+    a.href = url;
+    a.download = filename;
+    document.body.appendChild(a);
+    a.click();
+    setTimeout(function() {
+      document.body.removeChild(a);
+      window.URL.revokeObjectURL(url);
+    }, 0);
+  }
+};
+
+const toBool = (val) => {
+  return (val + '').toLowerCase() === 'true';
+};
+
+const addUser = (username, password, admin, vc, githubPAT) => {
+  let deferredObject = new $.Deferred();
+  userAuth.checkToken((token) => {
+    $.ajax({
+      url: `${webportalConfig.restServerUri}/api/v1/user`,
+      data: {
+        username,
+        password,
+        admin: toBool(admin),
+        modify: false,
+      },
+      type: 'PUT',
+      headers: {
+        Authorization: `Bearer ${token}`,
+      },
+      dataType: 'json',
+      success: () => {
+        let reqs = [];
+        if (githubPAT) {
+          let req = $.ajax({
+            url: `${webportalConfig.restServerUri}/api/v1/user/${username}/githubPAT`,
+            data: {
+              githubPAT: githubPAT,
+            },
+            type: 'PUT',
+            headers: {
+              Authorization: `Bearer ${token}`,
+            },
+            dataType: 'json',
+          });
+          reqs.push(req);
+        }
+        // Admin user VC update will be executed in rest-server
+        if (!toBool(admin) && vc) {
+          let req = $.ajax({
+            url: `${webportalConfig.restServerUri}/api/v1/user/${username}/virtualClusters`,
+            data: {
+              virtualClusters: vc,
+            },
+            type: 'PUT',
+            headers: {
+              Authorization: `Bearer ${token}`,
+            },
+            dataType: 'json',
+          });
+          reqs.push(req);
+        }
+        $.when(...reqs).then(
+          () => {
+            deferredObject.resolve(`User ${username} created successfully`);
+          },
+          (xhr) => {
+            const res = JSON.parse(xhr.responseText);
+            deferredObject.resolve(`User ${username} created successfully but failed when update other info: ${res.message}`);
+          }
+        );
+      },
+      error: (xhr, textStatus, error) => {
+        const res = JSON.parse(xhr.responseText);
+        deferredObject.resolve(`User ${username} created failed: ${res.message}`);
+      },
+    });
+  });
+  return deferredObject.promise();
+};
+
+const addUserRecursively = (userInfos, index, results) => {
+  if (index == 0) {
+    loading.showLoading();
+  }
+  if (index >= userInfos.length) {
+    loading.hideLoading();
+    alert(results.join('\r\n'));
```

Comment:
do not need `\r`

Reply:
Fixed.

---
In reply to: [272030346](https://github.com/Microsoft/pai/pull/2495#discussion_r272030346) [](ancestors = 272030346)